profiler/profile.go

```diff
@@ -236,22 +240,13 @@ func (l *limitedTraceCollector) Write(p []byte) (n int, err error) {
 		return
 	}
 	l.written += n
-	if l.written >= l.limit {
-		// We can't stop tracing within this method, because trace.Stop
-		// won't return until all pending events are written, which
-		// requires calling this function, meaning we'll deadlock.
-		go l.Stop()
+	if l.written >= l.limit && !l.stopped {
+		l.stopped = true
+		close(l.done)
```

Comment:
NIT: Could something like this be used to avoid the need for the `l.stopped` field? Not sure if it makes the code better, just an idea I had. Feel free to ignore.

```suggestion
	if l.written >= l.limit {
		select {
			case <-l.done:
			default:
				close(l.done)
		}
```

Reply:
I like that a lot better. Thanks!